Updated C# guidelines (#120)
IntelliTect/CodingGuidelines: docs/coding/csharp.md

# C# Coding Standards

## Sections
- [Naming](#naming)
  - [Abbreviations](#abbreviations)
  - [Casing](#casing)
  - [Namespaces](#namespaces)
  - [Types](#types)
  - [Methods](#methods)
  - [Variables and Fields](#variables-and-fields)
  - [Miscellaneous](#miscellaneous)

- [Layout](#layout)
  - [Comments](#comments)
  - [Files](#files)
  - [Whitespace](#whitespace)

- [Coding](#coding)
  - [Arrays](#arrays)
  - [Assemblies](#assemblies)
  - [Branches](#branches)
  - [Classes](#classes)
  - [Conversions](#conversions)

>   - [Dipsose()](#dipsose)

Reviewer: Typo?

Author: yep

  - [Synchronization](#synchronization)
  - [Tasks](#tasks)
  - [Threads](#threads)
  - [ToString()](#tostring)
  - [Types](#types)
  - [Miscellaneous](#miscellaneous)

## Naming

### Abbreviations
- :x: DO NOT: use abbreviations or contractions within identifier names.

### Casing
- :heavy_check_mark: DO: capitalize both characters in two-character acronyms, except for the first word of a camelCased identifier.
- :heavy_check_mark: DO: capitalize only the first character in acronyms with three or more characters, except for the first word of a camelCased identifier.
- :x: DO NOT: capitalize any of the characters in acronyms at the beginning of a camelCased identifier.
- :heavy_check_mark: DO: use PascalCasing for all class names.
- :heavy_check_mark: DO: use camelCasing for local variable names.
- :heavy_check_mark: DO: use uppercase literal suffixes (e.g., 1.618033988749895M).
- :heavy_check_mark: DO: use camelCasing for variable declarations using tuple syntax. 
- :grey_question: CONSIDER: using PascalCasing for all tuple item names.
- :heavy_check_mark: DO: use PascalCasing for namespace names.
- :heavy_check_mark: DO: use camelCasing for parameter names.

> - :grey_question: CONSIDER: using the same casing on a propertyâ€™s backing field as that used in the property, distinguishing the backing field with an â€œ_â€ prefix. Do not, however, use two underscores; identifiers beginning with two underscores are reserved for the use of t.

Reviewer 2: ```suggestion
- :grey_question: CONSIDER: using the same casing on a property's backing field as that used in the property, distinguishing the backing field with an "_" prefix. Do not, however, use two underscores; identifiers beginning with two underscores are reserved for the use of t.
```

> - :heavy_check_mark: CONSIDER: using the same casing on a propertyâ€™s backing field as that used in the property, distinguishing the backing field with an â€œ_â€ prefix. Do not, however, use two underscores; identifiers beginning with two underscores are reserved for the use of t.

Reviewer: Can we filter out the "smart" quotes that are coming from the CSV file? `with an â€œ_â€ prefix`

- :no_entry: AVOID: naming fields with camelCase.
- :heavy_check_mark: DO: name properties with PascalCase.

> - :heavy_check_mark: DO: use PascalCasing and an â€œIâ€ prefix for interface names.

Reviewer 2: ```suggestion
- :heavy_check_mark: DO: use PascalCasing and an "I" prefix for interface names.
```

### Types
- :x: DO NOT: define types without placing them into a namespace.
- :heavy_check_mark: DO: choose meaningful names for type parameters and prefix the name with T.
- :grey_question: CONSIDER: indicating a constraint in the name of a type parameter.
- :no_entry: AVOID: shadowing a type parameter of an outer type with an indentically named type parameter.

### Methods
- :heavy_check_mark: DO: give methods names that are verbs or verb phrases.  

### Variables and fields
- :heavy_check_mark: DO: favor clarity over brevity when naming identifiers.
- :x: DO NOT: use Hungarian notation (that is, do not encode the type of a variable in its name).
- :heavy_check_mark: DO: name classes with nouns or noun phrases.
- :heavy_check_mark: DO: use the C# keyword rather than the BCL name when specifying a data type (e.g., string rather than String).
- :x: DO NOT: use a constant for any value that can possibly change over time. The value of pi and the number of protons in an atom of gold are constants; the price of gold, the name of your company, and the version number of your program can change.
- :heavy_check_mark: DO: name properties using a noun, noun phrase, or adjective.

> - :heavy_check_mark: DO: favor prefixing Boolean properties with “Is,” “Can,” or “Has,” when that practice adds value.

Reviewer 2: ```suggestion
- :heavy_check_mark: DO: favor prefixing Boolean properties with "Is," "Can," or "Has," when that practice adds value.
```

- :heavy_check_mark: DO: declare all instance fields as private (and expose them via a property).
- :heavy_check_mark: DO: use constant fields for values that will never change.
- :no_entry: AVOID: constant fields for values that will change over time.
- :x: DO NOT: include “sentinel” values (such as a value called Maximum); such values can be confusing to the user.

### Miscellaneous
- :heavy_check_mark: DO: treat parameter names as part of the API, and avoid changing the names if version compatibility between APIs is important.
- :heavy_check_mark: DO: name the source file with the name of the public type it contains.
- :grey_question: CONSIDER: giving a property the same name as its type.
- :heavy_check_mark: DO: use the same name for constructor parameters (camelCase) and properties (PascalCase) if the constructor parameters are used to simply set the property.

> - :heavy_check_mark: DO: name an exception classes with the “Exception” suffix.

Reviewer 2: ```suggestion
- :heavy_check_mark: DO: name an exception classes with the "Exception" suffix.
```

- :heavy_check_mark: DO: ensure that Dispose() is idempotent (it should be possible to call Dispose() multiple times).
- :heavy_check_mark: DO: keep Dispose() simple, focusing on the resource cleanup required by finalization.
- :no_entry: AVOID: calling Dispose() on owned objects that have a finalizer. Instead, rely on the finalization queue to clean up the instance.
- :heavy_check_mark: DO: invoke a base class’s Dispose() method when overriding Dispose().
- :grey_question: CONSIDER: ensuring that an object becomes unusable after Dispose() is called. After an object has been disposed, methods other than Dispose() (which could potentially be called multiple times) should throw an ObjectDisposedException.
- :heavy_check_mark: DO: implement IDisposable on types that own disposable fields (or properties) and dispose of those instances.
- :heavy_check_mark: DO: invoke a base class’s Dispose() method from the Dispose(bool disposing) method if one exists.

### Enums
- :grey_question: CONSIDER: adding new members to existing enums, but keep in mind the compatibility risk.

> - :no_entry: AVOID: creating enums that represent an “incomplete” set of values, such as product version numbers.

Reviewer 2: ```suggestion
- :no_entry: AVOID: creating enums that represent an "incomplete" set of values, such as product version numbers.
```

> - :no_entry: AVOID: creating “reserved for future use” values in an enum.

Reviewer 2: ```suggestion
- :no_entry: AVOID: creating "reserved for future use" values in an enum.
```

- :no_entry: AVOID: enums that contain a single value.
- :heavy_check_mark: DO: provide a value of 0 (none) for simple enums, knowing that 0 will be the default value when no explicit initialization is provided.
- :heavy_check_mark: DO: use the FlagsAttribute to mark enums that contain flags.
- :heavy_check_mark: DO: provide a None value equal to 0 for all flag enums.

> - :no_entry: AVOID: creating flag enums where the zero value has a meaning other than “no flags are set.”

Reviewer 2: ```suggestion
- :no_entry: AVOID: creating flag enums where the zero value has a meaning other than "no flags are set."
```

- :no_entry: AVOID: general catch blocks and replace them with a catch of System.Exception.
- :no_entry: AVOID: catching exceptions for which the appropriate action is unknown. It is better to let an exception go unhandled than to handle it incorrectly.
- :no_entry: AVOID: catching and logging an exception before rethrowing it. Instead, allow the exception to escape until it can be handled appropriately.
- :heavy_check_mark: DO: prefer using an empty throw when catching and rethrowing an exception so as to preserve the call stack.
- :heavy_check_mark: DO: report execution failures by throwing exceptions rather than returning error codes.
- :x: DO NOT: have public members that return exceptions as return values or an out parameter. Throw exceptions to indicate errors; do not use them as return values to indicate errors.
- :x: DO NOT: use exceptions for handling normal, expected conditions; use them for exceptional, unexpected conditions.
- :no_entry: AVOID: throwing exceptions from property getters.
- :heavy_check_mark: DO: preserve the original property value if the property throws an exception.

> - :heavy_check_mark: DO: use nameof(value) (which resolves to “value”) for the paramName argument when creating ArgumentException() or ArgumentNullException() type exceptions (value" is the implicit name of the parameter on property setters)."

Reviewer 2: ```suggestion
- :heavy_check_mark: DO: use nameof(value) (which resolves to "value") for the paramName argument when creating ArgumentException() or ArgumentNullException() type exceptions (value" is the implicit name of the parameter on property setters)."
```
